### PR TITLE
dev/sg: generate named migrations and revert named migrations

### PR DESCRIPTION
**Pull request by bobheadxi**
Implements the 2nd and 3rd steps of https://github.com/sourcegraph/sourcegraph/issues/37244:

> - modify makeMigrationFilenames in makeMigrationFilenames to take a name as well as an index. Sanitize (s/ /_/g et al) the name and add it to the end of the created filenames
> - (a little less trivial) ensure the call to makeMigrationFilenames in the sg migration revert command receives the correct name so it can actually revert the correct filenames to modify

## Test plan

```
➜ go run ./dev/sg migration add --db codeintel 'do the thing!'
Migration files created
 Up query file: migrations/codeintel/1655353366_do_the_thing/up.sql
 Down query file: migrations/codeintel/1655353366_do_the_thing/down.sql
 Metadata file: migrations/codeintel/1655353366_do_the_thing/metadata.yaml
 
Checkout the development docs for migrations:  https://docs.sourcegraph.com/dev/background-information/sql/migrations
➜  git add .
➜  git ci "fake migration"
[06-14-dev/sg_generate_named_mgirations_and_revert_named_migrations 36ce124d9e] fake migration
 3 files changed, 14 insertions(+)
 create mode 100644 migrations/codeintel/1655353366_do_the_thing/down.sql
 create mode 100644 migrations/codeintel/1655353366_do_the_thing/metadata.yaml
 create mode 100644 migrations/codeintel/1655353366_do_the_thing/up.sql
➜  git hash                         
36ce124d9ee1b53e15fd5f6f50202a2c66807d60
➜  go run ./dev/sg migration revert 36ce124d9ee1b53e15fd5f6f50202a2c66807d60
Migration files redacted
 Up query file: /Users/robert@sourcegraph.com/Projects/sourcegraph/sourcegraph/migrations/codeintel/1655353366_do_the_thing/up.sql
 
Migration files created
 Up query file: migrations/codeintel/1655353400_revert_36ce124d9ee1b53e15fd5f6f50202a2c66807d60/up.sql
 Down query file: migrations/codeintel/1655353400_revert_36ce124d9ee1b53e15fd5f6f50202a2c66807d60/down.sql
 Metadata file: migrations/codeintel/1655353400_revert_36ce124d9ee1b53e15fd5f6f50202a2c66807d60/metadata.yaml
 
Checkout the development docs for migrations:  https://docs.sourcegraph.com/dev/background-information/sql/migrations
```

### CODE REVIEW

**Comment by bobheadxi** [comment]
Current dependencies on/for this PR:
* main
  * **PR #37247** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/37247" target="_blank">Graphite</a> 
    * **PR #37248** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/37248" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/37248?utm_source=stack-comment).

File: dev/sg/internal/migration/util_test.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/hexops/autogold"
+	"github.com/stretchr/testify/require"
+)
+
+func TestMakeMigrationFilenamesFromDir(t *testing.T) {
+	cases := []struct {
+		want          autogold.Value
+		migrationName string
+	}{
+		{
```
Comment (jhchabran) [inline dev/sg/internal/migration/util_test.go:15]: @bobheadxi If I add the following: 

```go
		{
			autogold.Want("basic filenames", MigrationFiles{
				UpFile:       "foobar/1/up.sql",
				DownFile:     "foobar/1/down.sql",
				MetadataFile: "foobar/1/metadata.yaml",
			}),
			"",
		},
```

The test will fail with:

```
--- FAIL: TestMakeMigrationFilenamesFromDir (0.17s)
    --- FAIL: TestMakeMigrationFilenamesFromDir/basic_filenames (0.17s)
        want.go:200: mismatch (-want +got):
            --- want
            +++ got
            @@ -1,4 +1,4 @@
             MigrationFiles{
            -   UpFile: "foobar/1/up.sql", DownFile: "foobar/1/down.sql",
            -   MetadataFile: "foobar/1/metadata.yaml",
            +   UpFile: "foobar/1_/up.sql", DownFile: "foobar/1_/down.sql",
            +   MetadataFile: "foobar/1_/metadata.yaml",
             }
            \ No newline at end of file
```

But `foobar/1/up.sql` is the expected path as per the existing migrations.

File: dev/sg/internal/migration/util.go [hunk 2]
```diff
@@ -29,16 +29,22 @@ type MigrationFiles struct {
...
-	downPath := filepath.Join(baseDir, fmt.Sprintf("%d/down.sql", migrationIndex))
-	metadataPath := filepath.Join(baseDir, fmt.Sprintf("%d/metadata.yaml", migrationIndex))
-	return MigrationFiles{upPath, downPath, metadataPath}, nil
+	return makeMigrationFilenamesFromDir(baseDir, migrationIndex, name)
+}
+
+func makeMigrationFilenamesFromDir(baseDir string, migrationIndex int, name string) (MigrationFiles, error) {
+	sanitizedName := strings.ReplaceAll(strings.ToLower(name), " ", "_")
+	return MigrationFiles{
+		UpFile:       filepath.Join(baseDir, fmt.Sprintf("%d_%s/up.sql", migrationIndex, sanitizedName)),
```
Comment (jhchabran) [inline dev/sg/internal/migration/util.go:44]: Isn't the mandatory `_` in `%d_%s` breaking the previous convention? When we didn't have a filename. See the other comment.

File: dev/sg/internal/migration/util_test.go (hunk 1 above)
Comment (bobheadxi) [reply dev/sg/internal/migration/util_test.go:15]: Good catch - I'll add a test for this case

File: dev/sg/internal/migration/util.go (hunk 2 above)
Comment (bobheadxi) [reply dev/sg/internal/migration/util.go]: Fixed in https://github.com/sourcegraph/sourcegraph/pull/37248/commits/88eb7542c16f72a3f5438855e27cf32eb21b7e9b !

File: dev/sg/internal/migration/util_test.go (hunk 1 above)
Comment (bobheadxi) [reply dev/sg/internal/migration/util_test.go:20]: Fixed in https://github.com/sourcegraph/sourcegraph/pull/37248/commits/88eb7542c16f72a3f5438855e27cf32eb21b7e9b !

**Review by jhchabran — APPROVED** [review]
🚀

File: dev/sg/internal/migration/util.go [hunk 3]
```diff
@@ -29,16 +30,34 @@ type MigrationFiles struct {
...
+}
+
+var nonAlphaNumericOrUnderscore = regexp.MustCompile("[^a-z0-9_]+")
+
+func makeMigrationFilenamesFromDir(baseDir string, migrationIndex int, name string) (MigrationFiles, error) {
+	sanitizedName := nonAlphaNumericOrUnderscore.ReplaceAllString(
+		strings.ReplaceAll(strings.ToLower(name), " ", "_"), "",
+	)
+	var dirName string
+	if sanitizedName == "" {
```
Comment (efritz) [inline dev/sg/internal/migration/util.go:49]: Does this condition happen inpractice?

**Pull request merged**